R/kenwardroger.R

```diff
@@ -0,0 +1,160 @@
+#' @param object `mmrm` object
+#' @param contrast `matrix` contrast matrix
+#' @param linear `logical`\cr whether to use linear Kenward-Roger approximation
+#' @return List with `num_df`, `denom_df`, `f_stat` and `p_val` (2-sided p-value).
+#' @keywords internal
+df_md_kr <- function(object, contrast, linear = FALSE) {
+  if (object$tmb_data$reml != 1) {
+    stop("Kenward-Roger is only for REML!")
+  }
+  kr_comp <- object$kr_comp
+  w <- solve(object$tmb_obj$he(object$theta_est))
+  v_adj <- object$beta_vcov_adj
+  df <- h_kr_df(object$beta_vcov, contrast, w, kr_comp$P)
+  vsolve <- solve(contrast %*% v_adj %*% t(contrast))
+  f_statistic <- 1 / nrow(contrast) * object$beta_est %*% t(contrast) %*% vsolve %*% contrast %*% object$beta_est
+  f_star <- f_statistic * df$lambda
+  ret <- list(
+    num_df = nrow(contrast),
+    num_df = nrow(contrast),
+    denom_df = df$m,
+    f_stat = f_star[1, 1],
+    p_val = pf(f_star[1, 1], nrow(contrast), df$m, lower.tail = FALSE)
+  )
+  ret$sqrt <- sqrt(contrast %*% v_adj %*% t(contrast))
+  ret
+}
+
+#' Calculation of Kenward-Roger Degrees of Freedom for One-Dimensional Contrast
+#' @param object `mmrm` object
+#' @param contrast contrast vector
+#' @param linear `logical`\cr whether to use linear Kenward-Roger approximation
+#' @return List with `est`, `se`, `df`, `t_stat` and `p_val`.
+#' @keywords internal
+df_1d_kr <- function(object, contrast, linear = FALSE) {
+  if (object$tmb_data$reml != 1) {
+    stop("Kenward-Roger is only for REML!")
+  }
+  assert_numeric(contrast, len = length(component(object, "beta_est")))
+  est <- sum(contrast * component(object, "beta_est"))
+  kr_comp <- object$kr_comp
+  w <- solve(object$tmb_obj$he(object$theta_est))
+  v_adj <- object$beta_vcov_adj
+  df <- h_kr_df(object$beta_vcov, matrix(contrast, nrow = 1), w, kr_comp$P)
+  se <- sqrt(contrast %*% v_adj %*% contrast)[1, 1]
+
+  ret <- list(
+    est = est,
+    se = se,
+    df = df$m
+  )
+  t_stat <- abs(est) / se
+  ret$t_stat <- t_stat
+  ret$p_val <- pt(t_stat, df$m, lower.tail = FALSE) / 2
+  ret
+}
+
+#' obtain the adjusted Kenward-Roger degree of freedom
+#' @param v0 unadjusted covariance matrix
+#' @param l linear combination matrix
+#' @param w hessian matrix
+#' @param p P matrix from `h_get_kr_comp`
+#' @return List with `m` and `lambda`
+#' @keywords internal
+h_kr_df <- function(v0, l, w, p) {
+  theta <- t(l) %*% solve(l %*% v0 %*% t(l)) %*% l
+  nl <- nrow(l)
+  thetav0 <- theta %*% v0
+  pl <- lapply(seq_len(nrow(p) / ncol(p)), function(x) {
+    ii <- (x - 1) * ncol(p) + 1
+    jj <- x * ncol(p)
+    p[ii:jj, ]
+  })
+  thetav0pv0 <- lapply(pl, function(x) {
+    thetav0 %*% x %*% v0
+  })
+  a1 <- 0
+  a2 <- 0
+  for (i in seq_len(length(pl))) {
+    for (j in seq_len(length(pl))) {
+      a1 <- a1 + w[i, j] * h_tr(thetav0pv0[[i]]) * h_tr(thetav0pv0[[j]])
+      a2 <- a2 + w[i, j] * h_tr(thetav0pv0[[i]] %*% thetav0pv0[[j]])
+    }
+  }
```

Comment:
also here would  add a comment line above this to help developers understand this, maybe link to formula in vignette.